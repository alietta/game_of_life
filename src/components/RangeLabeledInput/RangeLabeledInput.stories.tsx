import React from 'react';
import { DarkMode, LightMode, Layer } from 'sancho';
import { RangeLabeledInput } from './index';

export default { title: 'inputs' };

export const RangeStory = () => (
  <div css={{ width: 400 }}>
    <DarkMode>
      <Layer css={{ borderRadius: 0 }}>
        <RangeLabeledInput labelName="field size" />
      </Layer>
    </DarkMode>
    <LightMode>
      <Layer css={{ borderRadius: 0 }}>
        <RangeLabeledInput
          labelName="value"
          step={0.5}
          formValue={(value: string) => `${value} x ${value}`}
        />
      </Layer>
    </LightMode>
  </div>
);
