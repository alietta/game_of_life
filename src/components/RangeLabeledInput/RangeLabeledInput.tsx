import React, { FC, useState, ChangeEvent, useEffect, useCallback } from 'react';
import { Text, InputGroup } from 'sancho';

export type RangeLabeledInputProps = {
  formValue?: (value: string) => string;
  labelName: string;
  onValueChange?: (value: string) => void;
} & React.InputHTMLAttributes<HTMLInputElement>;

export const RangeLabeledInput: FC<RangeLabeledInputProps> = props => {
  const {
    formValue,
    labelName,
    onValueChange,
    max = 10,
    min = 1,
    step = 1,
    ...inputProps
  } = props;
  const getLabel = useCallback((val: string) => {
    const formatedValue = formValue ? formValue(val) : val;
    return `${labelName}: ${formatedValue}`;
  }, [labelName]);
  const [value, setValue] = useState<string>(`${max}`);
  const [label, setLabel] = useState<string>(getLabel(`${max}`));
  const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
    const { value: rangeValue } = event.target;
    setValue(event.target.value);
    onValueChange && onValueChange(rangeValue);
  };

  useEffect(() => {
    setLabel(getLabel(value));
  }, [value]);

  return (
    <InputGroup label={label} css={{ display: 'flex', flexDirection: 'column' }}>
      <input
        {...inputProps}
        placeholder="Name"
        type="range"
        min={min}
        max={max}
        step={step}
        value={value}
        onChange={handleChange}
      />
      <div css={{ display: 'flex', justifyContent: 'space-between', padding: '0 5px' }}>
        <Text>{min}</Text>
        <Text>{max}</Text>
      </div>
    </InputGroup>
  );
};
