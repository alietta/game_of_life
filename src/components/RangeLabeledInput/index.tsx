import React, { FC, useState, ChangeEvent, useEffect } from 'react';
import { useTheme, Text, InputGroup, Input } from 'sancho';

type Props = {
  formValue?: (value: string) => string;
  labelName: string;
  onValueChange?: (value: string) => void;
} & React.InputHTMLAttributes<HTMLInputElement>;

const RangeLabeledInput: FC<Props> = props => {
  const {
    formValue,
    labelName,
    onValueChange,
    max = 10,
    min = 1,
    step = 1,
    ...inputProps
  } = props;
  const [value, setValue] = useState<string>(`${min}`);
  const [label, setLabel] = useState<string>('');
  const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
    const { value } = event.currentTarget;
    setValue(value);
    onValueChange && onValueChange(value);
  };

  useEffect(() => {
    const formatedValue = formValue ? formValue(value) : value;
    setLabel(`${labelName}: ${formatedValue}`);
  }, [value, labelName]);

  return (
    <InputGroup label={label} css={{ display: 'flex', flexDirection: 'column' }}>
      <input
        {...inputProps}
        placeholder="Name"
        type="range"
        min={min}
        max={max}
        step={step}
        value={value}
        onChange={handleChange}
      />
      <div css={{ display: 'flex', justifyContent: 'space-between', padding: '0 5px' }}>
        <Text>{min}</Text>
        <Text>{max}</Text>
      </div>
    </InputGroup>
  );
};

export { RangeLabeledInput };
