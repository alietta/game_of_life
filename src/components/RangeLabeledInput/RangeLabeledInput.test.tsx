import React from 'react';
import { mount } from 'enzyme';
import { RangeLabeledInput } from './RangeLabeledInput';

describe('RangeLabeledInput', () => {
  it('test render', () => {
    const element = mount(<RangeLabeledInput labelName='values' />);
    expect(element).toMatchSnapshot();
  });
  it('should call onValueChange with props', () => {
    const onValueChange = jest.fn();
    const element = mount(<RangeLabeledInput labelName='values' onValueChange={onValueChange} />);
    element.find('input').simulate('change', {
      target: { value: '4' },
    });
    expect(onValueChange).toHaveBeenCalled();
    expect(onValueChange.mock.calls[0][0]).toBe('4');
  });
  it('should call formValue', () => {
    const formValue = jest.fn();
    const element = mount(<RangeLabeledInput labelName='values' formValue={formValue} />);
    element.find('input').simulate('change', {
      target: { value: '4' },
    });
    expect(formValue).toHaveBeenCalled();
  });
});
