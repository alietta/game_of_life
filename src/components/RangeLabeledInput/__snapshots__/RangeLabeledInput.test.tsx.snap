// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RangeLabeledInput test render 1`] = `
.emotion-6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-5 {
  margin-top: 1rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-5:first-child {
  margin-top: 0;
}

.emotion-1 {
  display: inline-block;
  margin-bottom: 0.5rem;
}

.emotion-0 {
  box-sizing: border-box;
  margin: 0;
  font-weight: 400;
  line-height: 1.5;
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,"Helvetica Neue",Arial,"Noto Sans",sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol","Noto Color Emoji";
  font-size: 1rem;
  color: #212529;
  -webkit-font-smoothing: antialiased;
  -webkit-text-size-adjust: none;
  font-size: 0.875rem;
  font-weight: 400;
  color: #5F6871;
  line-height: 1.5;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  padding: 0 5px;
}

.emotion-2 {
  box-sizing: border-box;
  margin: 0;
  font-weight: 400;
  line-height: 1.5;
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,"Helvetica Neue",Arial,"Noto Sans",sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol","Noto Color Emoji";
  font-size: 1rem;
  color: #212529;
  -webkit-font-smoothing: antialiased;
  -webkit-text-size-adjust: none;
  font-size: 1rem;
}

<RangeLabeledInput
  labelName="values"
>
  <Component
    className="emotion-6"
    label="values: 1"
  >
    <div
      className="emotion-5"
    >
      <Component
        htmlFor="1"
      >
        <label
          className="Label emotion-1"
          htmlFor="1"
        >
          <Component
            className="Label__text"
            variant="subtitle"
          >
            <div
              className="Label__text emotion-0"
            >
              values: 1
            </div>
          </Component>
        </label>
      </Component>
      <input
        max={10}
        min={1}
        onChange={[Function]}
        placeholder="Name"
        step={1}
        type="range"
        value="1"
      />
      <div
        className="emotion-4"
      >
        <Component>
          <span
            className="emotion-2"
          >
            1
          </span>
        </Component>
        <Component>
          <span
            className="emotion-2"
          >
            10
          </span>
        </Component>
      </div>
    </div>
  </Component>
</RangeLabeledInput>
`;
